ui/pages/asset/components/asset-chart.tsx

```diff
@@ -0,0 +1,349 @@
+      const index = Math.max(
+        0,
+        Math.min(
+          data.length - 1,
+          Math.round((chart.crosshairX / chart.width) * data.length),
+        ),
+      );
+
+      const x = xAxis.getPixelForValue(data[index].x);
+      const y = yAxis.getPixelForValue(data[index].y);
+
+      chart.ctx.lineWidth = 1;
+      chart.ctx.strokeStyle = '#BBC0C5';
+      chart.ctx.beginPath();
+      chart.ctx.moveTo(x, 0);
+      chart.ctx.lineTo(x, chart.height);
+      chart.ctx.stroke();
+
+      chart.ctx.beginPath();
+      chart.ctx.arc(x, y, 3, 0, 2 * Math.PI);
+      chart.ctx.fillStyle = chart.options.borderColor as string;
+      chart.ctx.fill();
+    }
+  },
+};
+
+Chart.register(
+  LinearScale,
+  PointElement,
+  LineElement,
+  Filler,
+  Decimation,
+  crosshairPlugin,
+);
+
+const initialChartOptions: ChartOptions<'line'> & { fill: boolean } = {
+  normalized: true,
+  parsing: false,
+  aspectRatio: 2.7,
+  layout: { autoPadding: false, padding: 0 },
+  animation: { duration: 0 },
+  fill: true,
+  backgroundColor: ({ chart }) => {
+    const gradient = chart.ctx.createLinearGradient(0, 0, 0, chart.height);
+    gradient.addColorStop(0, `${chart.options.borderColor}60`);
+    gradient.addColorStop(1, `${chart.options.borderColor}00`);
+    return gradient;
+  },
+  elements: {
+    line: { borderWidth: 2 },
+    point: { pointStyle: false },
+  },
+  plugins: {
+    decimation: {
+      algorithm: 'lttb',
+      samples: 150,
+      threshold: 150,
+      enabled: true,
+    },
+  },
+};
+
+// A chart showing historic prices for a native or token asset
+const AssetChart = ({
+  address,
+  currentPrice,
+}: {
+  address: string;
+  currentPrice?: number;
+}) => {
+  const t = useI18nContext();
+  const chainId = hexToDecimal(useSelector(getCurrentChainId));
+  const currency = useSelector(getCurrentCurrency);
+  const theme = useTheme();
+
+  const [data, setData] = useState<ChartData<'line', Point[]>>();
+  const [loading, setLoading] = useState<boolean>(true);
+  const [timeRange, setTimeRange] = useState<TimeRange>('1D');
+  const [chartOptions, setChartOptions] = useState(initialChartOptions);
+
+  const chartRef = useRef<Chart<'line', Point[]>>();
+  const priceRef = useRef<{
+    setPrices: (_: {
+      price?: number;
+      comparePrice?: number;
+      date?: number;
+    }) => void;
+  }>();
+  const maxPriceTooltip = useRef<{
+    setTooltip: (_: { xAxisPercent: number; price: number }) => void;
+  }>();
+  const minPriceTooltip = useRef<{
+    setTooltip: (_: { xAxisPercent: number; price: number }) => void;
+  }>();
+
+  useEffect(() => {
+    setLoading(true);
+
+    fetchWithCache({
+      url: `https://price-api.metafi.codefi.network/v1/chains/${chainId}/historical-prices/${address}?vsCurrency=${currency}&timePeriod=${timeRange}`,
+      cacheOptions: { cacheRefreshTime: MINUTE },
+      functionName: 'GetAssetHistoricalPrices',
+    })
+      .catch(() => ({}))
```

Comment:
i think that's better if we extract this to a separated hook ( can be called `useChartData.js` ).
this keeps your component focused on rendering and makes the data fetching logic reusable.